### ISSUE TITLE
Refreshing a page removes the cookie using the local provider

### ISSUE THREAD

**Issue by jorni-moddit** [sidebase/nuxt-auth#732]
### Environment

- Operating System: `Darwin`
- Node Version:     `v18.15.0`
- Nuxt Version:     `3.11.2`
- CLI Version:      `3.11.1`
- Nitro Version:    `2.9.6`
- Package Manager:  `npm@9.5.0`
- Builder:          `-`
- User Config:      `devtools`, `app`, `runtimeConfig`, `modules`, `auth`, `i18n`, `image`, `css`, `nitro`, `vite`, `routeRules`
- Runtime Modules:  `@nuxtjs/i18n@8.3.0`, `@nuxt/image@1.5.0`, `@vueuse/nuxt@10.9.0`, `@sidebase/nuxt-auth@0.7.1`
- Build Modules:    `-`

### Reproduction

We're using a WordPress back-end with the simple-JWT plugin. 
For this example I returned a hardcoded body in the api/auth/[...].ts
And there is a simple login form.

https://stackblitz.com/edit/github-gpkaqq?file=server%2Fapi%2Fauth%2F[...].ts

### Describe the bug

After logging in a cookie is set, when not configured it's under auth.token and I have an authenticated state.
![image](https://github.com/sidebase/nuxt-auth/assets/54063242/b1245554-e419-453f-b2ec-d327abaec522)

After a page refresh this cookie is removed.
![image](https://github.com/sidebase/nuxt-auth/assets/54063242/8591f945-cb99-4a97-922b-2606c50ef99c)

**Comment by husseinjahan** [comment]
I have the same Issue.

I'm using local authentication but connect to api to get user login. Sometimes not all the times when I refresh the browser, the response comes with a 'SET-COOKIE' response header that clears auth token (I don't know why!). I think there is a bug.

Is there any workaround about it?

![image](https://github.com/sidebase/nuxt-auth/assets/10027816/67959a57-2d8d-4dfc-806d-054f548be0d9)

**Comment by FireMasterK** [comment]
I'm facing the same issue, but the cookie isn't removed by HTTP header, but rather the client-side JS code.

**Comment by Norel** [comment]
I have the exact same problem except I can't reproduce the bug in a local project (dev or prod), only when deployed to a distant server.

**Comment by FireMasterK** [comment]
> I have the exact same problem except I can't reproduce the bug in a local project (dev or prod), only when deployed to a distant server.

Same case for me, except I'm able to reproduce it on the Stackblitz in production mode!

**Comment by phoenix-ru** [comment]
I can reproduce on `playground-local` using `pnpm build && pnpm nuxi preview`.

Adding this to the reproduction `app.vue` fixes the issue.
```ts
onMounted(() => {
  getSession({ force: true });
});
```

Why is it needed, however? And why doesn't it work without `force: true`?

I assume it is related to https://github.com/sidebase/nuxt-auth/blob/9bd9f452d29cf2b93c768c017702bc9719fb8adf/src/runtime/composables/local/useAuthState.ts#L52-L57

If someone could further investigate and propose a fix, it would be great!

**Comment by Norel** [comment]
It turns out that using `node 18` is problematic when using nuxt fetch (https://github.com/nuxt/nuxt/issues/12552).
Upgrading to `node 20` fixed the issue for me.
Stackblitz is probably using node 18 too.

**Comment by FireMasterK** [comment]
> It turns out that using `node 18` is problematic when using nuxt fetch ([nuxt/nuxt#12552](https://github.com/nuxt/nuxt/issues/12552)). Upgrading to `node 20` fixed the issue for me. Stackblitz is probably using node 18 too.

I'm able to reproduce it on node 20 still, my Dockerfile uses the node:20-slim image.

**Comment by ndro** [comment]
I have the same issue with local provider.

Cookie got removed in Firefox after reloading the browser or updating the app. But it's fine in Chrome.

my auth config in `nuxt.config.ts`:
```js
  auth: {
    baseURL: `${process.env.SOME_DOMAIN_API}`,
    provider: {
      type: 'local',
      pages: {
        login: '/login'
      },
      endpoints: {
        signIn: { path: '/login', method: 'post' },
        signOut: { path: '/logout', method: 'delete' },
        signUp: { path: '/register', method: 'post' },
        getSession: { path: '/me', method: 'get' }
      },
      token: {
        signInResponseTokenPointer: '/token',
        type: '',
        maxAgeInSeconds: 1209600
      },
      sessionDataType: {
        data: {
          // some custom data
        }
      },
    },
    globalAppMiddleware: {
      isEnabled: true
    }
  },
```

**Comment by albank** [comment]
My cookies are also set to empty 4 times like @husseinjahan above, when I refresh the page. I have the issue with both node18 and node20.
However, when using the network address created by nuxt, I don't have the issue:
Ex.
Network:  http://172.19.x.x:3000/

When I use my development mydomain.localhost that points to that address, my cookies get deleted 4x.

Any idea why ?

**Comment by mbellamyy** [comment]
Any updates?

**Comment by mbellamyy** [comment]
I can reproduce OP's scenario when I use Stackblitz's embedded browser (on the right column), but I cannot reproduce it when I open the output in a new tab

**Comment by mbellamyy** [comment]
> can reproduce on `playground-local` using `pnpm build && pnpm nuxi preview`.
> 
> Adding this to the reproduction `app.vue` fixes the issue.

It does indeed resolve it on client side, but in server side the user is still unauthenticated.

**Comment by GreenmeisterDavid** [comment]
I'm experiencing this issue as well; prod env, using local provider, the cookie disappears on reload on Firefox, while everything seems fine on Chromium-based browsers.

**Comment by taharehany** [comment]
After long and painstaking research, I found this: 

`const headers = useRequestHeaders(['cookie']) as HeadersInit
const { data } = await useFetch('/api/example', { headers })`

this will set cookies from the client side to the request you want in the Nuxt server side

**Comment by andre-silva9975** [comment]
I am having the same problem. I am also using local provider. Everything works in development and production mode while in localhost. But when I build the app in a remote server that serves as a reverse proxy (using nginx) between the client and the app running in the same server at localhost, every time I log in and reload the page, I am redirected to the login page after a few seconds and the access_token cookie disappears. Here is my configuration:

`auth: {
    baseURL: "/api/auth",
    globalAppMiddleware: true,
    provider: {
      type: "local",
      endpoints: {
        signIn: { path: "/login", method: "post" },
        signOut: { path: "/logout", method: "post" },
        getSession: { path: "/user", method: "get" }
      },
      pages: {
        login: "/login"
      },
      token: {
        signInResponseTokenPointer: "/token/accessToken",
        sameSiteAttribute: "lax",
      },
    },
  }`

**Comment by cip8** [comment]
I have the same issues while using the `refresh` scheme.
Tested it both with `0.7.2` and `0.8.0-alpha.2`.
Node version `20.11.1`
Nuxt `3.11` with Vue `3.4`

When using the refresh scheme the cookies are set. If I refresh the browser cookies are gone and I'm being redirected to the login page.

If I switch to `local` then everything works fine for me, this happens only for the `refresh` scheme.

My base config:
```  auth: {
    isEnabled: true,
    globalAppMiddleware: {
      isEnabled: true,
      allow404WithoutAuth: true,
      addDefaultCallbackUrl: true,
    },
    baseURL: '/',
    provider: {
      type: 'refresh',
      endpoints: {
        signIn: { path: '/api/auth/login', method: 'post' },
        signOut: { path: '/api/auth/logout', method: 'post' },
        refresh: { path: '/api/auth/refresh', method: 'post' },
        signUp: { path: '/api/auth/user', method: 'post' },
        getSession: { path: '/api/user/myself', method: 'get' },
      },
      pages: {
        login: '/',
      },
      token: { 
        signInResponseTokenPointer: '/access_token',
        maxAgeInSeconds: 60 * 15, // 15 min
        cookieName: 'access_token',
        sameSiteAttribute: 'lax',
      },
      refreshOnlyToken: false,
      refreshToken: { 
        signInResponseRefreshTokenPointer: '/refresh_token',
        refreshRequestTokenPointer: 'refresh_token',
        maxAgeInSeconds: 60 * 60 * 24 * 30, // 30 days
        cookieName: 'refresh_token',
      },
    },
    session: {
      enableRefreshPeriodically: false,
      enableRefreshOnWindowFocus: false,
    },
  },
```

**Please help** - any suggestion on how to fix or what's causing this is greatly appreciated 💙💙

**Comment by cip8** [comment]
If anyone else has the same issue as me, here's the fixed `plugins/refresh-token.server` file:

```
import { _fetch } from "../utils/fetch.mjs";
import { jsonPointerGet, objectFromJsonPointer, useTypedBackendConfig } from "../helpers.mjs";
import { defineNuxtPlugin, useAuthState, useRuntimeConfig } from "#imports";

export default defineNuxtPlugin({
  name: "refresh-token-plugin",
  enforce: "pre",
  async setup(nuxtApp) {
    const { rawToken, rawRefreshToken, refreshToken, token, lastRefreshedAt } = useAuthState();

    if (!refreshToken.value || !token.value) return;

    const config = nuxtApp.$config.public.auth;
    const configToken = useTypedBackendConfig(useRuntimeConfig(), "refresh");
    const { provider } = config;
    const { path, method } = provider.endpoints.refresh;

    try {
      let response = await _fetch(nuxtApp, path, {
        method,
        body: objectFromJsonPointer(provider.refreshToken.refreshRequestTokenPointer, refreshToken.value)
      });

      response = typeof response === 'string' ? JSON.parse(response) : response;

      const extractedToken = jsonPointerGet(response, provider.token.signInResponseTokenPointer);
      rawToken.value = extractedToken;

      if (!configToken.refreshOnlyToken) {
        const extractedRefreshToken = jsonPointerGet(response, provider.refreshToken.refreshRequestTokenPointer);
        rawRefreshToken.value = extractedRefreshToken;
      }

      lastRefreshedAt.value = new Date();
    } catch (err) {
      rawToken.value = null;
      rawRefreshToken.value = null;
    }
  }
});
```

This keeps me logged in even after refreshing the browser if the `refresh` scheme is used.

🖖💙

**Comment by andre-silva9975** [comment]
> I am having the same problem. I am also using local provider. Everything works in development and production mode while in localhost. But when I build the app in a remote server that serves as a reverse proxy (using nginx) between the client and the app running in the same server at localhost, every time I log in and reload the page, I am redirected to the login page after a few seconds and the access_token cookie disappears. Here is my configuration:
> 
> ` auth: {
> 
> ```
> baseURL: "/api/auth",
> globalAppMiddleware: true,
> provider: {
>   type: "local",
>   endpoints: {
>     signIn: { path: "/login", method: "post" },
>     signOut: { path: "/logout", method: "post" },
>     getSession: { path: "/user", method: "get" }
>   },
>   pages: {
>     login: "/login"
>   },
>   token: {
>     signInResponseTokenPointer: "/token/accessToken",
>     sameSiteAttribute: "lax",
>   },
> },
> ```
> 
> } `

I have updated to version [0.8.0-alpha.2](https://www.npmjs.com/package/@sidebase/nuxt-auth/v/0.8.0-alpha.2) and when I reload the page the user doesn't logged out.

**Comment by HELWATANY** [comment]
The issue is related to running your server (development server or on production) in a secure connection, trying the same steps while running localhost without `https` won't generate the issue, while setting the development server to run on `https` while cause the problem to occur.

### How To Run Your Local Development On HTTPS
1- You can follow this [article](https://medium.com/etc-a2z/setting-up-https-for-localhost-7e9a18253ca) to generate and set up your locally signed certificate.
2- Copy your certificates and key to `project-root-dir/certificates`
3- You can add the following environment variable in your `.env` file
```bash
NUXT_DEV_SERVER_CONNECTION_TYPE="https"  # Nuxt Development Server Type: Should be one of [http|https]
```
4- Update your `nuxt.config.ts` should look something like this
```typescript
export default defineNuxtConfig({
  devServer: {
    https:
      process.env.NUXT_DEV_SERVER_CONNECTION_TYPE === 'https'
        ? {
            key: './certificates/localhost.key',
            cert: './certificates/localhost.crt'
          }
        : false
  },
// ... the reset of your nuxt configs
})
```

It seems that the issue had been fixed starting [0.8.0-alpha.2](https://www.npmjs.com/package/@sidebase/nuxt-auth/v/0.8.0-alpha.2), I was able to run my development server with secure connection without facing this issue

**Comment by fontanes-innatial** [comment]
> The issue is related to running your server (development server or on production) in a secure connection, trying the same steps while running localhost without `https` won't generate the issue, while setting the development server to run on `https` will cause the problem to occur.
> 
> ### How To Run Your Local Development On HTTPS
> 1- You can follow this [article](https://medium.com/etc-a2z/setting-up-https-for-localhost-7e9a18253ca) to generate and set up your locally signed certificate. 2- Copy your certificates and key to `project-root-dir/certificates` 3- You can add the following environment variable in your `.env` file
> 
> ```shell
> NUXT_DEV_SERVER_CONNECTION_TYPE="https"  # Nuxt Development Server Connection Type: Should be one of [http|https]
> ```
> 
> 4- Update your `nuxt.config.ts` should look something like this
> 
> ```ts
> export default defineNuxtConfig({
>   devServer: {
>     https:
>       process.env.NUXT_DEV_SERVER_CONNECTION_TYPE === 'https'
>         ? {
>             key: './certificates/localhost.key',
>             cert: './certificates/localhost.crt'
>           }
>         : false
>   },
> // ... the reset of your nuxt configs
> })
> ```
> 
> It seems that the issue had been fixed starting [0.8.0-alpha.2](https://www.npmjs.com/package/@sidebase/nuxt-auth/v/0.8.0-alpha.2), I was able to run my development server with secure connection without facing this issue

I tried the partner's solution but it didn't work in my case.

**Comment by yamachita0109** [comment]
I tried the same but it didn't work.

```package.json
    "@sidebase/nuxt-auth": "^0.8.0-alpha.2",
```

```nuxt.config.ts
    auth: {
        baseURL: process.env.API_BASE_URL,
        globalAppMiddleware: true,
        provider: {
            type: 'local',
            pages: {
                login: "/login"
            },
            endpoints: {
                signIn: { path: "admin/login", method: "post" },
                signOut: { path: "admin/logout", method: "post" },
                signUp: { path: "admin/login", method: "post" },
                getSession: { path: "admin/user", method: "get" },
            },
            token: {
                signInResponseTokenPointer: "/token",
                type: "Bearer",
                headerName: "Authorization",
                maxAgeInSeconds: 60 * 60 * 24,
                sameSiteAttribute: "lax",
            },
        }
    },
```

**Comment by alfredomtzrmz** [comment]
I'm facing the same issue, when log in, the cookie set correctly then refresh page and the cookie gets removed. 

- I'm using external backend
- I set **sameSiteAttribute** to **strict**

```json
"@sidebase/nuxt-auth": "^0.8.0"
```

```ts
  auth: {
    baseURL: 'http://external_api.com/api/auth/',
    globalAppMiddleware: true,
    provider: {
      type: 'refresh',
      endpoints: {
        signIn: { path: 'login', method: 'post' },
        signOut: { path: 'logout', method: 'post' },
        signUp: false,
        getSession: { path: 'me', method: 'get' },
        refresh: { path: 'refresh', method: 'post' },
      },
      token: {
        sameSiteAttribute: 'strict',
        signInResponseTokenPointer: '/data/access_token',
        cookieName: 'access_token',
        maxAgeInSeconds: 60 * 60,
      },
      refreshToken: {
        signInResponseRefreshTokenPointer: '/data/access_token',
        cookieName: 'access_token',
        refreshRequestTokenPointer: 'access_token',
        maxAgeInSeconds: 20160 * 60,
      },
      pages: {
        login: '/login',
      },
    },
  },
  ```

**Comment by Ulrich-Mbouna** [comment]
Same problem

**Comment by Ulrich-Mbouna** [comment]
Any Updates ? Please i face this issue

**Comment by mhamzamughal86** [comment]
It seems that the issue you're facing with the cookie being removed upon page refresh is due to the SSR (Server-Side Rendering) nature of the application.

To resolve this issue, you can disable SSR for your application by setting ssr: false in the nuxt.config.ts file. This will make your application run as a Single Page Application (SPA), ensuring that the cookie is available and preserved across page refreshes.

`export default defineNuxtConfig({
  ssr: false,
  // other configurations
})
`

**Comment by Ulrich-Mbouna** [comment]
> It seems that the issue you're facing with the cookie being removed upon page refresh is due to the SSR (Server-Side Rendering) nature of the application.
> 
> To resolve this issue, you can disable SSR for your application by setting ssr: false in the nuxt.config.ts file. This will make your application run as a Single Page Application (SPA), ensuring that the cookie is available and preserved across page refreshes.
> 
> `export default defineNuxtConfig({ ssr: false, // other configurations }) `

I tried it and it solve the issue, but i'm asking myself if it's not a bad idea to disable ssr in nuxt application. 🤔

**Comment by HELWATANY** [comment]
> ```ts
> maxAgeInSeconds: 
> ```

This is can not be considered a solution if your main goal is to create an SSR application!

**Comment by HELWATANY** [comment]
@yamachita0109 @mtzrmzia If you are serving your app over a secure connection, please consider setting `secureCookieAttribute` to `true`, please check [here](https://auth.sidebase.io/guide/local/quick-start#securecookieattribute)

```typescript
  auth: {
    baseURL: 'http://external_api.com/api/auth/',
    globalAppMiddleware: true,
    provider: {
      type: 'refresh',
      endpoints: {
        signIn: { path: 'login', method: 'post' },
        signOut: { path: 'logout', method: 'post' },
        signUp: false,
        getSession: { path: 'me', method: 'get' },
        refresh: { path: 'refresh', method: 'post' },
      },
      token: {
        sameSiteAttribute: 'strict',
        signInResponseTokenPointer: '/data/access_token',
        cookieName: 'access_token',
        maxAgeInSeconds: 60 * 60,
        secureCookieAttribute: true, // <---- you need to add this, default is false and should be false if you are serving over HTTP
      },
      refreshToken: {
        signInResponseRefreshTokenPointer: '/data/access_token',
        cookieName: 'access_token',
        refreshRequestTokenPointer: 'access_token',
        maxAgeInSeconds: 20160 * 60,
        secureCookieAttribute: true, // <---- you need to add this, default is false and should be false if you are serving over HTTP
      },
      pages: {
        login: '/login',
      },
    },
  },
```

/sugges

**Comment by 5mind** [comment]
Any update on this please ? Did anyone figure it out ?

**Comment by vincepaquette** [comment]
Any update on this please ? Did anyone figure it out ? Disabling the SSR isn't a solution in my case

**Comment by cip8** [comment]
> Any update on this please ? Did anyone figure it out ? Disabling the SSR isn't a solution in my case

Have you tried the new version ([0.9.1](https://github.com/sidebase/nuxt-auth/releases/tag/0.9.1)) & experiencing the same issues?

**Comment by HELWATANY** [comment]
> > Any update on this please ? Did anyone figure it out ? Disabling the SSR isn't a solution in my case
> 
> Have you tried the new version ([0.9.1](https://github.com/sidebase/nuxt-auth/releases/tag/0.9.1)) & experiencing the same issues?

@cip8 
I wasn't able to regenerate the issue since version `0.8.0-alpha.2`

**Comment by Ulrich-Mbouna** [comment]
> > > Any update on this please ? Did anyone figure it out ? Disabling the SSR isn't a solution in my case
> > 
> > 
> > Have you tried the new version ([0.9.1](https://github.com/sidebase/nuxt-auth/releases/tag/0.9.1)) & experiencing the same issues?
> 
> @cip8 I wasn't able to regenerate the issue since version `0.8.0-alpha.2`

So your problem is solved?

**Comment by HELWATANY** [comment]
> > > > Any update on this please ? Did anyone figure it out ? Disabling the SSR isn't a solution in my case
> > > 
> > > 
> > > Have you tried the new version ([0.9.1](https://github.com/sidebase/nuxt-auth/releases/tag/0.9.1)) & experiencing the same issues?
> > 
> > 
> > @cip8 I wasn't able to regenerate the issue since version `0.8.0-alpha.2`
> 
> So your problem is solved?

I was trying to help by providing reproduction steps and a fix if needed 🙂

**Comment by josselinonduty** [comment]
I still get this issue on v0.9.1.

I have tried a few different setups, like using h3 as my session provider, and using nuxt-auth endpoints/token. Sometimes it does work, but it seems not to be reliable.

I'll edit my message in a few hours and provide my config or a reproduction if I can.

**Comment by josselinonduty** [comment]
> I still get this issue on v0.9.1.
> 
> I have tried a few different setups, like using h3 as my session provider, and using nuxt-auth endpoints/token. Sometimes it does work, but it seems not to be reliable.
> 
> ~~I'll edit my message in a few hours and provide my config or a reproduction if I can.~~

Okay, so I got it working.
Here is what I found:

*Before*:
```ts
export default defineEventHandler(async (event) => {
    const session = getCookie(event, "session");
    // business logic here
});
```
The issue is that the cookie is not sent reliably or there is something I don't understand.
What I can confirm is that I get session `undefined` when I reload the page manually (hence it invalidates the client cookie because throw an error when it's invalid).

*After*:
```ts
export default defineEventHandler(async (event) => {
    const authorization = event.headers.get("Authorization");
    const token = authorization?.split("Bearer ")[1];
});
```
Now I get the token reliably and can use it.


*Conclusions*:
There might be an issue (or a conflicting intended behaviour) in Nuxt(-Auth) lifecycle that doesn't pass cookies or not at the right time, hence it is `undefined`. I could be totally wrong tho.

**Comment by samuelemusiani** [comment]
> It seems that the issue you're facing with the cookie being removed upon page refresh is due to the SSR (Server-Side Rendering) nature of the application.
> 
> To resolve this issue, you can disable SSR for your application by setting ssr: false in the nuxt.config.ts file. This will make your application run as a Single Page Application (SPA), ensuring that the cookie is available and preserved across page refreshes.
> 
> `export default defineNuxtConfig({ ssr: false, // other configurations }) `

I have encountered this issue too. Disabling ssr fixed for me

**Comment by dasmoro** [comment]
In my case changing `maxAgeInSeconds` of token from 30 to 21 days was fixed such problem. With 30 days value cookie set only with 1 second expire. I don't know, why. And didn't find any documentation of this.
![Скриншот 11-10-2024 142433](https://github.com/user-attachments/assets/bbafcf0a-f37f-4d11-8712-0935144ed7f8)

**Comment by rmnaderdev** [comment]
Hi all. I also wanted to chime in and say that I too am facing this issue on v0.9.3. I am in the middle of moving a Vue app to Nuxt in order to use SSR to support SEO better for my product so turning SSR off is not an acceptable fix in my situation. I am trying to use my existing .NET backend auth with the new Nuxt frontend (JWT tokens with refresh tokens). I am able to login and make API calls but, like others have said, the auth cookie is removed once I refresh the page with the following console error. 

![image](https://github.com/user-attachments/assets/37380fde-028b-4b56-bc7d-8a53ffb121ea)
![image](https://github.com/user-attachments/assets/5449a60f-dc52-4462-bbb8-d8495181a69a)

I thought maybe the issue was cors or an https issue, but I have already turned on... 

Okay so as I am typing out this, I realized that I did not add the `NODE_TLS_REJECT_UNAUTHORIZED=0` variable to my .env file like I did with my other proof-of-concept project and the reason SSR was not able to get the user was because my .NET self-signed SSL cert was not trusted by the node backend when calling the endpoints on the server side...

So it appears that I am no longer having this issue. I hope this helps someone else!

**Comment by ZenelShabani** [comment]
God this is frustrating, can someone plx look into this. Im getting the same issue on 0.9.3.

Cookie is being set, and then deleted as soon as one refreshes.

**Comment by mikolajszymczuk1** [comment]
Any new state of this bug ?

**Comment by mikolajszymczuk1** [comment]
I have simple solution , you can enable option (`disableServerSideAuth: true`)

**Comment by julienguillot77** [comment]
I'm facing the same issue so I tried your solution but when I set the config to `disableServerSideAuth: true`, I have infinite redirect loop from '/session' to '/auth/login'.
Btw, if you don't create a /session route, you got an error when accessing to 'auth/login' page.

**Comment by julienguillot77** [comment]
It seems that set the token cookie's httpOnlyCookieAttribute value to false makes persisting the cookie even after reloading.
When set to true, the cookie is kind of hidden or stored somewhere else. I would like to make this cookie httpOnly. Is a fix could possibly be made ?

**Comment by mikolajszymczuk1** [comment]
@julienguillot77 let me see but what I observed for example when I set cookie domain in nuxt settings , cookie exists after refresh

**Comment by wpoPR** [comment]
any news about this?

**Comment by dolphinwow** [comment]
I have the same problem. Is it still not fixed?

**Comment by mikolajszymczuk1** [comment]
@dolphinwow there are not new updates, im waiting with my auth refactor but i think i will kick this lib to trash and implement own auth logic.

**Comment by mikolajszymczuk1** [comment]
last information that i have is that this is not problem with cookie , i changed config, i added few options from nuxt auth docs and finally cookie exists after refresh page. After some exploration in source code i found there is a strict problem with getting session after refresh, its something like logic correctly load data to variables but there is error when lib makes request

**Comment by mikolajszymczuk1** [comment]
later i will setup this repo and i will try to debug this error, I have some idea when can be the problem

**Comment by uricholiveira** [comment]
Any news? Still having this problem using version: 0.9.4

**Comment by cyprille** [comment]
Hi everyone and many thanks for this very cool library!

I'm experiencing the exact same issue as everyone one here since August. I struggled hours/days to make it work, but without any luck, trying all updates since then...

As a turnaround, I disabled SSR, but I'm hoping this could be fixed some time to work out of the box, as it's a big downside to be forced to disable SSR.

If I can help to resolve it, feel free to contact me 🙏🏻

**Comment by mikolajszymczuk1** [comment]
@cyprille hi , if you whould like to help , you can help me with debugging, try to setup project and what i know is the problem exists on external api, i tested everything on playground and with api created on nitro side session works correcltry but when logic must refresh session after refresh page on our external api there is error in request call function. Im working on this bug today, this is terrible but if we fix it it will not force the most developers to switch to own auth logic

**Comment by cyprille** [comment]
@mikolajszymczuk1 Hi, thank you for your reply! Yes, I already have a project setup to reproduce this issue, but it's a private repo, so I won't be able to share it. 

It's easy to reproduce it with an external backend for auth with this kind of config: 

`auth: {
    isEnabled: true,
    baseURL: '/api-back',
    provider: {
      type: 'local',
      endpoints: {
        signIn: {
          path: '/security/login',
          method: 'post',
        },
        getSession: {
          path: '/backoffice-users/session',
          method: 'get',
        },
      },
      session: {
        dataResponsePointer: '/',
        dataType: {
          id: 'number',
          firstname: 'string',
          lastname: 'string',
          email: 'string',
        },
      },
      token: {
        signInResponseTokenPointer: '/token',
        type: 'Bearer',
        cookieName: 'auth.token',
        headerName: 'Authorization',
        maxAgeInSeconds: 50400,
        sameSiteAttribute: 'none',
        secureCookieAttribute: true,
        httpOnlyCookieAttribute: false,
      },
      refresh: {
        isEnabled: true,
        endpoint: {
          path: '/security/refresh',
          method: 'post',
        },
        refreshOnlyToken: true,
        token: {
          signInResponseRefreshTokenPointer: '/refresh_token',
          refreshRequestTokenPointer: '/refresh_token',
          cookieName: 'auth.refresh_token',
          maxAgeInSeconds: 50400,
          sameSiteAttribute: 'none',
          secureCookieAttribute: true,
          httpOnlyCookieAttribute: false,
        },
      },
    },
    globalAppMiddleware: true,
  },`

As you can see, I tried to set up a lot of configs here, following the documentation.
I also tried to restart from basic config to this point without any luck, testing each new parameter individually.

Hope this helps to debug!

**Comment by kaidolex** [comment]
I also experienced this issue on production but not during development, what I did is I added a domain on cookie in my nest backend code

Something like this:
![Image](https://github.com/user-attachments/assets/4b2dc806-a26a-41fc-8090-1e93b64efb67)

After adding domain my cookies are not being deleted after refresh

my domain setup is 
api.mydomain.com = nest backend
app.mydomain.com = nuxt frontend

**Comment by cyprille** [comment]
Hi everyone!

After upgrading to the latest version 0.10.0, I updated the configurations according to the documentation.
I added the new `cookieDomain` param and re-enabled the `ssr` to true, and now it works as expected 🎉 🔥

Here is my full auth configuration:

`auth: {
    baseURL: process.env.BACKEND_URL,
    provider: {
      type: 'local',
      endpoints: {
        signIn: {
          path: '/login',
          method: 'post',
        },
        getSession: {
          path: '/session',
          method: 'get',
        },
      },
      session: {
        dataResponsePointer: '/',
        dataType: {
          id: 'number',
          firstname: 'string',
          lastname: 'string',
          email: 'string',
        },
      },
      token: {
        signInResponseTokenPointer: '/token',
        type: 'Bearer',
        cookieName: 'auth.token',
        cookieDomain: process.env.APP_DOMAIN,
        headerName: 'Authorization',
        maxAgeInSeconds: 50400,
        sameSiteAttribute: 'lax',
        secureCookieAttribute: true,
        httpOnlyCookieAttribute: false,
      },
      refresh: {
        isEnabled: true,
        endpoint: {
          path: '/refresh',
          method: 'post',
        },
        refreshOnlyToken: true,
        token: {
          signInResponseRefreshTokenPointer: '/refresh_token',
          refreshResponseTokenPointer: '/token',
          refreshRequestTokenPointer: '/refresh_token',
          cookieName: 'auth.refresh_token',
          cookieDomain: process.env.APP_DOMAIN,
          maxAgeInSeconds: 50400,
          sameSiteAttribute: 'lax',
          secureCookieAttribute: true,
          httpOnlyCookieAttribute: false,
        },
      },
    },
    globalAppMiddleware: true,
  },`

Hope this helps 🙂

**Comment by joaoaugustogrobe** [comment]
### TL;DR: 
```ts
sameSiteAttribute: 'lax'
````

After hours of debugging this issue, I will post here some discoveries.

First, some context:
I have a external API that handles authentication using JWT.
The authentication flow was working, but when refreshing the page, the authentication state was lost.

I tried to play with different domains, adding a local domain (host file on mac), set cookies in the login response, and more... Nothing worked.

I started looking in the implementation, and here is a intersting piece of code:
```ts
  const _rawTokenCookie = useCookie(config.token.cookieName, {
    default: () => null,
    domain: config.token.cookieDomain,
    maxAge: config.token.maxAgeInSeconds,
    sameSite: config.token.sameSiteAttribute,
    secure: config.token.secureCookieAttribute,
    httpOnly: config.token.httpOnlyCookieAttribute
  });
  const rawToken = useState("auth:raw-token", () => _rawTokenCookie.value);
  watch(rawToken, () => {
    _rawTokenCookie.value = rawToken.value;
  });
```
rawToken is set during the signIn method, and there is a useCookie ref watching it, and storing it to a cookie.

Then I made some tests with the useCookie to understand the behavior:
```ts
const tokenCookie = useCookie('foo');
tokenCookie.value = 'bar';
```
Inspecting the browser, I could see the foo:bar cookie.

Adding the options one by one, everything was working until I added the `sameSiteAttribute: 'none'`. The cookie was not being set. After reading more about the sameSiteAttribute, as far as could understand, [browser will reject SameSite=None if the secure flag is not present](https://chromestatus.com/feature/5633521622188032), which was my case. In my mind, setting to 'none' would make the cookie more "insecure", and easier to setup in dev mode.

<details>
<summary>Same Site Attributes</summary>

- true will set the SameSite attribute to Strict for strict same site enforcement.

- false will not set the SameSite attribute.

- 'lax' will set the SameSite attribute to Lax for lax same site enforcement.

- 'strict' will set the SameSite attribute to Strict for strict same site enforcement.

- 'none' will set the SameSite attribute to None for an explicit cross-site cookie.
</details>

Setting the `sameSiteAttribute: 'lax'` worked with secure and httpOnly set to false.

This exact config may not work for every case, but the idea is to check if the config that you're using really makes the cookie accessible to your page.

Here is my working config:
```ts
      token: {
        signInResponseTokenPointer: '/accessToken',
        maxAgeInSeconds: 3600,
        // type: 'Bearer', Default value
        // cookieName: 'auth.token', Default value
        // headerName: 'Authorization', Default value
        // sameSiteAttribute: 'lax', // Default value
        // cookieDomain: '', // TODO: change to the domain
        // secureCookieAttribute: false, // Default value
        // httpOnlyCookieAttribute: false, // Default value
      },
```

**Comment by visualnaut** [comment]
I try every solution on this issue comments but still have the issue except by disabling SSR but unfortunately it's not an option for my case. Is anyone still have this issue as I do?

I'm using `"@sidebase/nuxt-auth": "^0.10.1",` at the moment, my config:
```js
{
  globalAppMiddleware: true,
  originEnvKey: 'NUXT_PUBLIC_API_BASE',
  provider: {
    type: 'local' as const,
    pages: {
      login: '/',
    },
    endpoints: {
      signIn: { path: '/go-agent/login', method: 'post' },
      getSession: { path: '/agent/v2/user/complete-profile', method: 'get' },
      signOut: false,
      signUp: false,
    },
    session: {
      dataResponsePointer: '/data',
    },
    token: {
      signInResponseTokenPointer: '/data/token',
      headerName: 'Authorization',
      type: 'Bearer',
    },
    refresh: {
      isEnabled: true,
      endpoint: { path: '/agent/v2/sessions/v2/refresh', method: 'post' },
      token: {
        signInResponseRefreshTokenPointer: '/data/refresh',
        refreshResponseTokenPointer: '/data/token',
        refreshRequestTokenPointer: '/refresh',
      }
    },
  },
}
``` 

Anyone have an alternatives on implementing local provider with refresh like this package?

**Comment by Ari24-cb24** [comment]
Currently experiencing the same issue, tried most of the solutions here 
```
token: {
        signInResponseTokenPointer: "/token",
        type: "Bearer",
        cookieName: "auth.token",
        headerName: "Authorization",
        maxAgeInSeconds: 86400,
        sameSiteAttribute: "lax",
        cookieDomain: "localhost",
        secureCookieAttribute: false,
        httpOnlyCookieAttribute: false
      },
```

Only working fix right now is disabling `ssr` which is fine at the moment but not sustainable for the future.

**Comment by phoenix-ru** [comment]
If someone is still having this issue, it would be great if you could try out the following PR:
https://github.com/sidebase/nuxt-auth/pull/1033

It attempts to fix loading state in the Nuxt middleware

```sh
npm i https://pkg.pr.new/@sidebase/nuxt-auth@1033
```

**Comment by cyprille** [comment]
Hi everyone!

Has anyone tested the PR from @phoenix-ru?
Any idea about this PR to be included in an upcoming nuxt-auth version?

Many thanks!

**Comment by phoenix-ru** [comment]
Hi @cyprille , you could give it a try yourself and document the outcome (what was before and what is after), this would be appreciated. Otherwise we would not proceed with merging the PR since we've got no acknowledgments of the problem being fixed by #1033
